src/sinks/prometheus/exporter.rs

```diff
@@ -160,39 +179,160 @@ impl SinkConfig for PrometheusCompatConfig {
 struct PrometheusExporter {
     server_shutdown_trigger: Option<Trigger>,
     config: PrometheusExporterConfig,
-    metrics: Arc<RwLock<ExpiringMetrics>>,
+    metrics: Arc<RwLock<IndexMap<MetricRef, (Metric, MetricMetadata)>>>,
     acker: Acker,
 }
 
-struct ExpiringMetrics {
-    map: IndexMap<MetricEntry, MetricMetadata>,
-    last_flush_timestamp: i64,
+/// Expiration metadata for a metric.
+#[derive(Clone, Copy, Debug)]
+struct MetricMetadata {
+    should_clear: bool,
+    last_updated: Instant,
 }
 
-struct MetricMetadata {
-    is_incremental_set: bool,
-    updated_at: Instant,
+impl MetricMetadata {
+    pub fn new(should_clear: bool) -> Self {
+        Self {
+            should_clear,
+            last_updated: Instant::now(),
+        }
+    }
+
+    pub fn has_expired(&self, timeout: Duration) -> bool {
+        self.last_updated.elapsed() > timeout
```

Comment:
I recommend avoiding using `.elapsed` in loop contexts, as it issues a `gettimeofday` (or equivalent) syscall via `Instant::now` on each call. You could pass in `last_flush` when it is called below, and possibly pre-calculate the expiry time instead of storing the last update and subtracting.

Reply:
It's worth pointing out that on the vast majority of systems, these clock calls (`clock_gettime`, in this case, since we're dealing with `Instant` and not `SystemTime`) will be vDSO accelerated: 15-25 nanosecond per call, more or less.

That said, though, your suggestion is still a viable lil' performance improvement that's easy to do, so I'll do that.

Reply:
Hmm, criterion says otherwise, despite the cycle counts being close:
```
rust-timeit --use 'std::time::Instant' --setup 'let ref1 = Instant::now(); let ref2 = Instant::now();' 'ref1.elapsed()' 'ref2.duration_since(ref1)'

`ref1.elapsed()`        time:   [1.1380 us 1.1407 us 1.1436 us]                              
`ref2.duration_since(ref1)`                                                                             
                        time:   [4.1161 ns 4.1651 ns 4.2423 ns]

`ref1.elapsed()`        time:   [3615.5721 cycles 3621.3010 cycles 3627.0069 cycles]         
`ref2.duration_since(ref1)`                                                                             
                        time:   [11.9903 cycles 12.0031 cycles 12.0168 cycles]

```

Reply:
This is where it can be tricky, on a per-system basis, but on my local Ubuntu server:

```
toby@consigliere:~$ cat /sys/devices/system/clocksource/clocksource0/current_clocksource
tsc

toby@consigliere:~$ rust-timeit --use 'std::time::Instant' --setup 'let ref1 = Instant::now(); let ref2 = Instant::now();' 'ref1.elapsed()' 'ref2.duration_since(ref1)'
...

`ref1.elapsed()`            time:   [31.101 ns 31.151 ns 31.193 ns]
`ref2.duration_since(ref1)` time:   [4.1393 ns 4.1411 ns 4.1434 ns]
```

Overall, yeah, there are definitely cases where it can be unexpectedly slow.  My comment wasn't meant to be a catch-all, but on a production server, I would expect the kernel tunables to be set such that the results on that server match what I saw running the above code.

Said another way: we shouldn't _always_ be scared of time, but to your point, it doesn't hurt to try and amortize the cost where it can be done without affecting the output/behavior.

Reply:
Hmm, interesting, my system is using `hpet` instead of the TSC, probably because the TSC was marked as unstable.

Reply:
Yeah, it's tricky.  This is one of the main reasons I wrote `quanta`. 😂